### PR TITLE
Release Test suites

### PR DESCRIPTION
**Pull request by ryanjduffy**
> [!Warning]
> * We should not squash + merge this branch so it retains the commit history from the feature branch

Merges the test suite dashboard feature branch into `main`

### CODE REVIEW

**Comment by ryanjduffy** [comment]
@bvaughn - this is ready for your review now. I've called out a few points of interest outside of the `TestRuns` and `Tests` changes. The Typecheck CI failure is a regression from https://github.com/replayio/devtools/pull/10198.

**Comment by bvaughn** [comment]
@ryanjduffy Got it. That regression has been fixed by @hbenl so this PR can be rebased/updated whenever.

**Comment by hbenl** [comment]
> a squashed commit we could revert in case we have a regression

If we create a merge commit (instead of "Rebase and merge"), reverting should still be possible, though I've never done this.

**Comment by bvaughn** [comment]
Cool, let's do a freeze on this branch for now. I'm a couple of hours into a many hours long review and it would just help me a bit if we locked things down until I leave my first round of feedback :)

**Comment by bvaughn** [comment]
In hindsight, I would have preferred _not_ moving components out of the "new" folder until after the merge because it's adding a lot of noise.

File: src/ui/components/Library/Team/View/TestRuns/TestRunListItem.module.css [hunk 1]
```diff
@@ -0,0 +1,3 @@
+.testsuitesSuccess {
+  background-color: var(--testsuites-v2-success-icon);
+}
```
Comment (bvaughn) [inline src/ui/components/Library/Team/View/TestRuns/TestRunListItem.module.css:3]: This file is not referenced by anything...
```suggestion
```

File: src/ui/components/Library/Team/View/TestRuns/TestRunDetailsPanel/RunStats.module.css [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+  padding: 0 0.5rem;
+}
+
+.flakyPill {
+  color: var(--theme-base-100);
+  background-color: var(--testsuites-flaky-color);
+}
+
+.failedPill {
+  color: #fff; /* we want white in both themes, or else the contrast is bad */
```
Comment (bvaughn) [inline src/ui/components/Library/Team/View/TestRuns/TestRunDetailsPanel/RunStats.module.css:19]: This should still be a style variable.

File: packages/shared/graphql/types.ts [hunk 3]
```diff
@@ -225,6 +225,9 @@ export interface Recording {
   userId?: string;
   userRole?: RecordingRole;
   workspace?: Workspace;
+  testRun?: {
+    id: string;
+  } | null;
```
Comment (bvaughn) [inline packages/shared/graphql/types.ts:230]: nit: This would be easier to work with as a string id:
```suggestion
  testRunId?: string;
```

File: packages/e2e-tests/helpers/authentication.ts [hunk 4]
```diff
@@ -11,14 +11,13 @@ export const E2E_USER_1_API_KEY = "ruk_jukvxbSz7syp4Tw21RzEwSK2bjucNDklCEDmVkjbH
...
 // Can view the "Replay Devtools Snapshots" (Test Suites) workspace
 // but cannot upload new recordings nor source-maps
 // trunk-ignore(gitleaks/generic-api-key)
 export const E2E_USER_4_API_KEY = "rwk_CaQCN8h7Wz794K1Yx2200FOcta24j7NkJaYO07TNndV";
 export const E2E_USER_4_TEAM_ID = "dzpmMDZkNWI5OS05MWYzLTRhYTktYTYyNC0zMDRjYjJlYjBlYzk=";
+
+// With only read:workspace role
+// trunk-ignore(gitleaks/generic-api-key)
+export const TEST_RUN_WORKSPACE_API_KEY = "rwk_ONquDEbGGXrjs0yvuON1BhSiAUdlAz8CAl7BlB4igHo";
+export const TEST_RUN_WORKSPACE_TEAM_ID = "dzplODhmZjY1Mi0wYTdmLTRhZDgtOTNiYS04OTZmMTBjYjRjM2Q=";
```
Comment (bvaughn) [inline packages/e2e-tests/helpers/authentication.ts:23]: Why did these change? Should this change be merged to main?

**Review by bvaughn — COMMENTED** [review]
This PR is too big. My brain is fried. Here's some feedback.

File: src/ui/components/Library/Team/View/Tests/hooks/useSyncTestIdToUrl.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  // Sync URL changes to component state
+  useEffect(() => {
+    const testIdFromState = prevTestIdRef.current;
+    const { testOrTestRunId: testIdFromUrl } = parseQueryParams(router.query);
+    if (testIdFromUrl) {
+      if (testIdFromState !== testIdFromUrl) {
+        setTestId(testIdFromUrl);
+      }
+    }
+  }, [router.query, setTestId]);
```
[the remark this answers is not in the document]
Comment (hbenl) [reply src/ui/components/Library/Team/View/Tests/hooks/useSyncTestIdToUrl.ts:39]: I think the right way would be to have this state only in the URL without duplicating it in `TestContext`. In other words `TestContext` would need to be changed so that its `testId` comes from the URL and its `setTestId()` writes directly to the URL.

File: src/ui/components/Library/Team/View/shared/Alert.tsx [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    "No replays were found for this run. They may be outside the retention window or may not have been uploaded",
+  MISSING_REPLAYS_FOR_TEST:
+    "This test has executions that are not reflected in the list of replays below",
+};
+
+export function Alert({ reason, link }: { reason: keyof typeof AlertList; link?: string }) {
+  return (
+    <div data-test-id={reason} className="gap-3 rounded-lg bg-chrome p-3">
+      <span>{AlertList[reason]}</span>
+      <a href={link} rel="noreferrer" target="_blank" className="ml-1 underline">
```
Comment (hbenl) [inline src/ui/components/Library/Team/View/shared/Alert.tsx:12]: We should not show "Learn more" if we don't have a `link`.

**Review by hbenl — APPROVED** [review]

File: src/ui/components/Library/Team/View/Tests/hooks/useSyncTestIdToUrl.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/Tests/hooks/useSyncTestIdToUrl.ts:39]: Agreed. This will be removed by SCS-1793

File: src/ui/components/Library/Team/View/TestRuns/hooks/useSyncTestStateToUrl.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      router.query
+    );
+    if (testRunIdFromUrl) {
+      if (testRunIdFromState !== testRunIdFromUrl) {
+        setTestRunId(testRunIdFromUrl);
+      }
+      if (testIdFromState !== testIdFromUrl && setTestId) {
+        setTestId(testIdFromUrl);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/TestRuns/hooks/useSyncTestStateToUrl.ts:49]: Going away with SCS-1793

File: src/ui/components/Library/Team/View/Tests/hooks/useSyncTestIdToUrl.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bvaughn) [reply src/ui/components/Library/Team/View/Tests/hooks/useSyncTestIdToUrl.ts:39]: I didn't realize there were pending bug-fixes for merge into test-suites. Seems like we're reviewing this prematurely.

**Comment by bvaughn** [comment]
@ryanjduffy Ping me again once #10158 or #10159 have landed and there are no more PRs pending that aren't included here. Feels kind of inefficient to spend time uncovering bugs in this PR that are already known.

**Comment by ryanjduffy** [comment]
> @ryanjduffy Ping me again once #10158 or #10159 have landed and there are no more PRs pending that aren't included here. Feels kind of inefficient to spend time uncovering bugs in this PR that are already known.

I did not intend to land #10158 on `test-suites` to avoid unnecessary changes to this branch but we can do that if you'd rather.

**Comment by bvaughn** [comment]
> but we can do that if you'd rather.

I think it would make the most sense for us to be reviewing as close to a final, bug-free thing as we can with this PR to main.

File: src/ui/components/Library/Team/View/Tests/TestListPanel/TestListItem.tsx [hunk 8]
```diff
@@ -1,32 +1,45 @@
...
+function Status({ test }: { test: Test }) {
+  const { sortBy } = useContext(TestContext);
 
-  if (failureRate > 0) {
+  const { status, rate, classNames } = useMemo(() => {
+    if (sortBy === "flakyRate") {
+      return { status: "flaky", rate: test.flakyRate, classNames: styles.flakyPill };
+    } else if (test.failureRate > 0) {
+      // sortBy could be "failureRate" or "alphabetical", show failure rate in that case
+      return { status: "failure", rate: test.failureRate, classNames: styles.failedPill };
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/Tests/TestListPanel/TestListItem.tsx:19]: This was discussed a bit in SCS-1805. We decided to leave as is for now.

File: src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts [hunk 9]
```diff
@@ -1,25 +1,43 @@
...
-  return value;
+
+  useEffect(() => {
+    if (status === STATUS_PENDING) {
+      // To avoid resetting the test runs to an empty array when loading
+      return;
+    }
+
+    setTestRuns(value);
+  }, [value, status]);
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts:38]: It's being used to retain the data from previous date range while the next range is loading. Is there another way we should be achieving that?

File: package.json [hunk 10]
```diff
@@ -29,6 +29,7 @@
     "@lexical/link": "^0.6.5",
     "@lexical/mark": "^0.6.5",
     "@lexical/react": "^0.6.5",
+    "@nivo/line": "^0.84.0",
```
[the remark this answers is not in the document]
Comment (bvaughn) [reply package.json:32]: I'm replacing this component over on #10211

File: src/ui/components/Library/Team/View/Tests/TestDetailsPanel/TestDetails.tsx [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  } else if (result === "flaky") {
+    if (total === 1) {
+      // If there's only one replay but the test is flaky,
+      // it's cypress and the flake was resolved internally
+      return "flaky";
+    } else {
+      // return passing for the first (last) one, flaky for the rest
+      return index === 0 ? "passed" : "flaky";
+    }
+  } else {
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/Tests/TestDetailsPanel/TestDetails.tsx:57]: This is caused by a gap in the graphql API. We are returning the computed result for test in a run but not results for each execution of the test in that run.

In other words, we return "flaky" for a test run and the recordings we have for that test but not "here's execution 1 which failed and made 2 replays" and "here's execution 2 which passed and made 2 replays". The fix for this is the scope of SCS-1843.

[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/Tests/TestDetailsPanel/TestDetails.tsx:57]: I didn't really answer your question ... sorry ... this code is a workaround for the gap I described. It's typically correct for most scenarios but seems to be falling down in the multiple replay case you mentioned here.

File: src/ui/components/Library/Team/View/TestRuns/TestRunDetailsPanel/RunStats.module.css (hunk 2 above)
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/TestRuns/TestRunDetailsPanel/RunStats.module.css:19]: @jonbell-lot23 - do we have an existing variable for this?

File: packages/e2e-tests/helpers/authentication.ts (hunk 4 above)
Comment (ryanjduffy) [reply packages/e2e-tests/helpers/authentication.ts:23]: These didn't change. They are new values for the test suite e2e tests

File: src/ui/components/Library/Team/View/TestRuns/TestRunListPanel/Chart.tsx [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+  return [
+    {
+      id: "Failure rate (%)",
+      color: "var(--testsuites-graph-gradient-stroke)",
+      data: sortedData,
+    },
+  ];
+}
+
+export const Chart = () => {
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/TestRuns/TestRunListPanel/Chart.tsx:63]: Resolved by #10211 (Thanks, @bvaughn !)

File: packages/e2e-tests/authenticated/tests-suite/test-runs-01.test.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+
+  //#region >>> When all tests pass should display a green checkmark to the left of the test run title
+  const testRunItemIcon = testRun.locator('[data-test-status="success"]');
+  expect(await testRunItemIcon.count()).toBe(1);
+  //#endregion
+
+  //#region >>> The relative time of the test run creation date should be displayed to the right of the test run title
+  const testRunItemDate = testRun.locator('[data-test-id="TestRun-Date"]');
+  expect(await testRunItemDate.count()).toBe(1);
+  //#endregion
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/e2e-tests/authenticated/tests-suite/test-runs-01.test.ts:45]: Correct. We're asserting that the test run row contains a particular child element.

File: packages/e2e-tests/README.md [hunk 14]
```diff
@@ -124,6 +124,7 @@ Here we're running the `breakpoints-05 test`, which depends on the `doc_debugger
 ```AUTOMATED_TEST_SECRET=<find me at backend/src/graphql-api/schema.ts!> \
 GRAPHQL_ADDRESS=http://localhost:8087/v1/graphql \
 AUTHENTICATED_TESTS_WORKSPACE_API_KEY=$RECORD_REPLAY_API_KEY \
+GOLDEN_TEST_RUN_WORKSPACE_API_KEY=<workspace api key to a test workspace>
```
Comment (ryanjduffy) [inline packages/e2e-tests/README.md:127]: TODO: Add content describing how these new tests work

File: packages/e2e-tests/authenticated/tests-suite/test-runs-01.test.ts [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+  noTestRunSelectedMessage,
+  noTestRunsMessage,
+  noTestSelected,
+  testItems,
+  testRunResult,
+  testRunSummary,
+  testRunsItems,
+} from "./test-suite.utils";
+
+test.use({ testRunState: "SUCCESS_IN_MAIN_WITH_SOURCE" });
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/e2e-tests/authenticated/tests-suite/test-runs-01.test.ts:22]: Renaming these to be more clear. I've also added a description in the first file that clarifies why these data objects are used and what the recording IDs mean. I'll add similar details to the README.

File: packages/e2e-tests/authenticated/tests-suite/test-suite.utils.ts [hunk 16]
```diff
@@ -0,0 +1,101 @@
+import { Locator, Page } from "@playwright/test";
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/e2e-tests/authenticated/tests-suite/test-suite.utils.ts:1]: Moved to helpers and renamed to clarify its for the dashboard (vs the devtools panels)

File: .env.sample [hunk 17]
```diff
@@ -17,3 +17,4 @@ FORM_CARRY_TOKEN=
 # API keys used by e2e tests
 AUTOMATED_TEST_SECRET=
 AUTHENTICATED_TESTS_WORKSPACE_API_KEY=
+GOLDEN_TEST_RUN_WORKSPACE_API_KEY=
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply .env.sample:20]: I added a block to README that better describes how these new tests work and why they all use the same recording id.

File: packages/e2e-tests/authenticated/tests-suite/test-runs-01.test.ts [hunk 18]
```diff
@@ -0,0 +1,143 @@
...
+
+  //#region >>> Filter test by text
+  await filterSummaryTestsByText(page, clientKey);
+  expect(await testItems(page).count()).toBe(2);
+  expect(await testItems(page).nth(0).innerText()).toContain(clientKey);
+  expect(await testItems(page).nth(1).innerText()).toContain(clientKey);
+  await filterSummaryTestsByText(page, "");
+  //#endregion
+
+  //#region >>> Filter by status
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply packages/e2e-tests/authenticated/tests-suite/test-runs-01.test.ts]: I kept the comments which help add some waypoints on what is being tested but removed the region tags.

File: src/ui/components/Library/Team/View/TestRuns/TestRunPanelWrapper.tsx [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+        setDeferredPending(true);
+      }, 200);
+
+      return () => {
+        clearTimeout(timer);
+      };
+    } else {
+      setDeferredPending(false);
+    }
+  }, [isPending]);
```
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/TestRuns/TestRunPanelWrapper.tsx:26]: This is removed by #10158

File: src/ui/components/Library/Team/View/TestRuns/TestRunDetailsPanel/RunStats.module.css (hunk 2 above)
Comment (jonbell-lot23) [reply src/ui/components/Library/Team/View/TestRuns/TestRunDetailsPanel/RunStats.module.css]: @ryanjduffy color: var(--theme-base-100);  should work, and it's what we use in the other .failedPill locations

File: src/ui/components/Library/Team/View/shared/Alert.tsx (hunk 6 above)
Comment (bvaughn) [reply src/ui/components/Library/Team/View/shared/Alert.tsx:12]: Good call. I missed this!

File: src/ui/components/Library/Team/utils.ts [hunk 20]
```diff
@@ -6,9 +6,11 @@ import { getRecordingWorkspace } from "ui/reducers/app";
 import { useAppSelector } from "ui/setup/hooks";
 
 export function parseQueryParams(query: ParsedUrlQuery) {
-  const [teamId, view, testRunId] = Array.isArray(query.param) ? query.param : [query.param!];
+  const [teamId, view, testOrTestRunId, _, testId] = Array.isArray(query.param)
+    ? query.param
+    : [query.param!];
 
-  return { teamId, testRunId: testRunId || null, view: view as View };
+  return { teamId, testOrTestRunId: testOrTestRunId || null, view: view as View, testId };
```
[the remark this answers is not in the document]
Comment (bvaughn) [reply src/ui/components/Library/Team/utils.ts]: This got marked as resolved but it doesn't seem to have changed (at least not in the way I suggested). WE're still returning an ambiguous `testOrTestRunId` value, now _in addition to_ a `testRun` value.

File: src/ui/components/Library/Team/View/TestRuns/TestRunTestPanel/TestRunTestPanel.tsx [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+  const { testRunId, testRuns } = useContext(TestRunsContext);
+
+  const { groupedTests, tests } = useTestRunDetailsSuspends(testRunId);
+  const selectedSpecTests =
+    // Select tests that not filtered in second panel
+    tests
+      ?.filter(
+        t => filterTestsByText === "" || t.sourcePath.toLowerCase().includes(filterTestsByText)
+      )
+      ?.filter(t => t.testId === testId) ?? [];
```
Comment (bvaughn) [inline src/ui/components/Library/Team/View/TestRuns/TestRunTestPanel/TestRunTestPanel.tsx:24]: I was a little confused by the naming here, given "tests" and "test id" seems to imply (to me) that there would only ever be one match. I wonder if there's a clearer name we could pick for this?

That being said, when I see chained `.filter` statements, I usually think "we could write this to be more efficient" using a single loop:
```suggestion
tests?.filter((test) => {
  if (test.testId !== testId) {
    return false;
  } else if (filterTestsByText !== "") {
    if (!test.sourcePath.toLowerCase().includes(filterTestsByText)) {
      return false;
    }
  }
  return true;
}) ?? [];
```

File: src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (bvaughn) [reply src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts:38]: I see. I think this is because we're not using Suspense, so we don't get that (the is-pending, dimmed behavior) for free.

Can we fetch this data using suspense (e.g. cache.read) rather than the imperative hook?

I don't know if there's any gotchas in this, but it seems like it would be nice to remove `testRuns` from the context , and just load it inside of `TestRunsList` or `TestRunDetailsPanel` using Suspense like we do with the other data (aka rename this hook to `useTestRunsSuspends`)

File: packages/e2e-tests/README.md [hunk 22]
```diff
@@ -149,12 +150,19 @@ You'll also probably need to specifically share that recording as "Public", espe
...
 
+### Test Suite Dashboard Tests
+
+The test suite dashboard tests (within `/test-suite-dashboard`) validate how the dashboard renders in various data scenarios created by the test runner plugins. Each scenario is modeled by an object in `./helpers/setupTestRun.ts` that represents different test run results a plugin might produce. Those objects are added to the database using the same methods the plugin would use and are retrieved by the dashboard to assert on expected behavior. We ensure multiple tests can run at the same time by generating a new UUID for the `clientKey` for the test run scenario used and looking for that UUID in the results.
+
+> _Note:_ The scenarios include the same recording ID for all recordings. That recording is never accessed by the tests but must be a valid, non-deleted recording for the backend to return the expected data in the dashboard. The recorded does not need to be re-recorded like other examples but can be updated to a new valid recording ID if necessary.
+
```
Comment (bvaughn) [inline packages/e2e-tests/README.md:158]: Nice 👍🏼 Thanks

**Review by bvaughn — COMMENTED** [review]
This looks really close. Two small requests.

File: src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (bvaughn) [reply src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts:38]: Cool 👍🏼  I'm having a bit of trouble keeping up with what's the current state here given the size of the PR

File: src/ui/components/Library/Team/View/shared/Alert.tsx (hunk 6 above)
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/shared/Alert.tsx:12]: `link` is provided everywhere this is used so I will mark `link` as required. I think it makes sense to always have additional docs for these warnings so this will enforce that.

File: packages/e2e-tests/examples.json [hunk 23]
```diff
@@ -5,54 +5,54 @@
     "requiresManualUpdate": true
   },
   "authenticated_comments.html": {
-    "recording": "b50dd5b7-ab70-4987-8987-cc3240dd2cf3",
-    "buildId": "macOS-chromium-20240110-e5625ebd3edc-50d01be708a2"
+    "recording": "94775d4e-33e7-4c02-8622-3b212503ebab",
+    "buildId": "linux-chromium-20240123-609ec6caa105-50d01be708a2"
```
Comment (bvaughn) [inline packages/e2e-tests/examples.json:9]: I think we should revert the changes tot his file. They aren't really a part of this change set, are they?

File: packages/e2e-tests/scripts/loadRecording.ts [hunk 24]
```diff
@@ -29,9 +29,11 @@ export async function loadRecording(recordingId: string) {
...
-  });
-
-  completeLog();
+  try {
+    await client.sendCommand("Recording.processRecording", {
+      recordingId,
+    });
+  } finally {
+    completeLog();
+  }
```
Comment (bvaughn) [inline packages/e2e-tests/scripts/loadRecording.ts:38]: Why is this showing up in the DIFF view? I'm confused.

**Comment by bvaughn** [comment]
Something looks funky here.

**Comment by ryanjduffy** [comment]
> Something looks funky here. I think you need to rebase or merge this branch.

Yeah, there were some odd changes from a previous merge. I merged again and all the spurious changes seem gone.

Now that the back button PR is merged into this and I've addressed your last 2 requests, i'm tagging you for (hopefully) a final ✅

File: src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (ryanjduffy) [reply src/ui/components/Library/Team/View/TestRuns/hooks/useTestRuns.ts]: No worries. I was waiting to tag you until it had landed but you got ahead of me :)

**Review by bvaughn — APPROVED** [review]
I scanned the most recent changes and I think things look okay. My brain is kind of 💀  at this point because the PR is so large, so 🤞🏼
